File: client/static/js/src/__tests__/metadata-conformance.test.ts
import './setup';
import { afterAll, beforeAll, describe, expect, test } from 'bun:test';
import fixture from '../../../../../crypto/testdata/crypto-conformance-v2.json';
import {
  AAD_FIELD_FILENAME,
  AAD_FIELD_PASSWORD_HINT,
  AAD_FIELD_SHA256,
  AAD_FIELD_TAGS,
} from '../crypto/aad';
import { fromHex, toBase64, toHex } from '../crypto/primitives';

const originalFetch = globalThis.fetch;
const CHUNKING_CONFIG = {
  plaintextChunkSizeBytes: 16777216,
  envelope: {
    version: 2,
    headerSizeBytes: 35,
    saltSizeBytes: 32,
    kdfProfile: 1,
    keyTypes: { account: 1, custom: 2 },
  },
  aesGcm: {
    nonceSizeBytes: 12,
    tagSizeBytes: 16,
    keySizeBytes: 32,
  },
};

beforeAll(() => {
  (globalThis as any).fetch = async (url: string | URL | Request) => {
    const urlStr = typeof url === 'string' ? url : url instanceof URL ? url.href : url.url;
    if (urlStr.includes('/api/config/chunking')) {
      return new Response(JSON.stringify(CHUNKING_CONFIG), {
        status: 200,
        headers: { 'Content-Type': 'application/json' },
      });
    }
    return originalFetch(url as any);
  };
});
afterAll(() => {
  globalThis.fetch = originalFetch;
});

import { decryptMetadataField } from '../crypto/metadata-helpers';

describe('owner metadata shared fixture', () => {
  const fields = [
    {
      name: 'filename',
      field: AAD_FIELD_FILENAME,
      plaintext: fixture.owner_metadata.filename_plaintext,
      nonceHex: fixture.owner_metadata.filename_nonce_hex,
      cipherHex: fixture.owner_metadata.filename_ciphertext_and_tag_hex,
    },
    {
      name: 'sha256',
      field: AAD_FIELD_SHA256,
      plaintext: fixture.owner_metadata.sha256_plaintext,
      nonceHex: fixture.owner_metadata.sha256_nonce_hex,
      cipherHex: fixture.owner_metadata.sha256_ciphertext_and_tag_hex,
    },
    {
      name: 'password_hint',
      field: AAD_FIELD_PASSWORD_HINT,
      plaintext: fixture.owner_metadata.password_hint_plaintext,
      nonceHex: fixture.owner_metadata.password_hint_nonce_hex,
      cipherHex: fixture.owner_metadata.password_hint_ciphertext_and_tag_hex,
    },
    {
      name: 'tags',
      field: AAD_FIELD_TAGS,
      plaintext: fixture.owner_metadata.tags_plaintext,
      nonceHex: fixture.owner_metadata.tags_nonce_hex,
      cipherHex: fixture.owner_metadata.tags_ciphertext_and_tag_hex,
    },
  ] as const;

  test('decrypts pinned filename, hash, hint, and tags ciphertext', async () => {
    const key = fromHex(fixture.password_kdf.account_key_hex);
    for (const field of fields) {
      const nonce = fromHex(field.nonceHex);
      const ciphertext = fromHex(field.cipherHex);
      const got = await decryptMetadataField(
        toBase64(ciphertext),
        toBase64(nonce),
        key,
        fixture.file_id,
        field.field,
        fixture.owner_username,
      );
      expect(got).toBe(field.plaintext);
    }
  });

  test('re-encrypts to the pinned ciphertext', async () => {
    const keyBytes = fromHex(fixture.password_kdf.account_key_hex);
    const key = await crypto.subtle.importKey('raw', keyBytes, 'AES-GCM', false, ['encrypt']);
    const aadByField: Record<string, string> = {
      [AAD_FIELD_FILENAME]: fixture.aad.encrypted_filename_hex,
      [AAD_FIELD_SHA256]: fixture.aad.encrypted_sha256sum_hex,
      [AAD_FIELD_PASSWORD_HINT]: fixture.aad.encrypted_password_hint_hex,
      [AAD_FIELD_TAGS]: fixture.aad.encrypted_tags_hex,
    };
    for (const field of fields) {
      const nonce = fromHex(field.nonceHex);
      const aad = fromHex(aadByField[field.field]);
      const encrypted = new Uint8Array(await crypto.subtle.encrypt({
        name: 'AES-GCM',
        iv: nonce,
        additionalData: aad,
        tagLength: 128,
      }, key, new TextEncoder().encode(field.plaintext)));
      expect(toHex(encrypted)).toBe(field.cipherHex);
    }
  });
});

File: crypto/metadata_conformance_test.go
package crypto

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"encoding/hex"
	"encoding/json"
	"os"
	"testing"
)

type ownerMetadataFixture struct {
	FilenamePlaintext     string `json:"filename_plaintext"`
	FilenameNonceHex      string `json:"filename_nonce_hex"`
	FilenameCipherHex     string `json:"filename_ciphertext_and_tag_hex"`
	SHA256Plaintext       string `json:"sha256_plaintext"`
	SHA256NonceHex        string `json:"sha256_nonce_hex"`
	SHA256CipherHex       string `json:"sha256_ciphertext_and_tag_hex"`
	PasswordHintPlaintext string `json:"password_hint_plaintext"`
	PasswordHintNonceHex  string `json:"password_hint_nonce_hex"`
	PasswordHintCipherHex string `json:"password_hint_ciphertext_and_tag_hex"`
	TagsPlaintext         string `json:"tags_plaintext"`
	TagsNonceHex          string `json:"tags_nonce_hex"`
	TagsCipherHex         string `json:"tags_ciphertext_and_tag_hex"`
}

func TestMetadataFieldSharedFixture(t *testing.T) {
	raw, err := os.ReadFile("testdata/crypto-conformance-v2.json")
	if err != nil {
		t.Fatal(err)
	}
	var corpus struct {
		FileID        string `json:"file_id"`
		OwnerUsername string `json:"owner_username"`
		PasswordKDF   struct {
			AccountKeyHex string `json:"account_key_hex"`
		} `json:"password_kdf"`
		AAD struct {
			Filename     string `json:"encrypted_filename_hex"`
			SHA256       string `json:"encrypted_sha256sum_hex"`
			PasswordHint string `json:"encrypted_password_hint_hex"`
			Tags         string `json:"encrypted_tags_hex"`
		} `json:"aad"`
		OwnerMetadata ownerMetadataFixture `json:"owner_metadata"`
	}
	if err := json.Unmarshal(raw, &corpus); err != nil {
		t.Fatal(err)
	}
	key, err := hex.DecodeString(corpus.PasswordKDF.AccountKeyHex)
	if err != nil {
		t.Fatal(err)
	}

	cases := []struct {
		name      string
		field     string
		plaintext string
		nonceHex  string
		cipherHex string
		aadHex    string
	}{
		{"filename", AADFieldFilename, corpus.OwnerMetadata.FilenamePlaintext, corpus.OwnerMetadata.FilenameNonceHex, corpus.OwnerMetadata.FilenameCipherHex, corpus.AAD.Filename},
		{"sha256", AADFieldSha256, corpus.OwnerMetadata.SHA256Plaintext, corpus.OwnerMetadata.SHA256NonceHex, corpus.OwnerMetadata.SHA256CipherHex, corpus.AAD.SHA256},
		{"password_hint", AADFieldPasswordHint, corpus.OwnerMetadata.PasswordHintPlaintext, corpus.OwnerMetadata.PasswordHintNonceHex, corpus.OwnerMetadata.PasswordHintCipherHex, corpus.AAD.PasswordHint},
		{"tags", AADFieldTags, corpus.OwnerMetadata.TagsPlaintext, corpus.OwnerMetadata.TagsNonceHex, corpus.OwnerMetadata.TagsCipherHex, corpus.AAD.Tags},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			nonce, err := hex.DecodeString(tc.nonceHex)
			if err != nil {
				t.Fatal(err)
			}
			wantCT, err := hex.DecodeString(tc.cipherHex)
			if err != nil {
				t.Fatal(err)
			}
			aad := BuildMetadataFieldAAD(corpus.FileID, tc.field, corpus.OwnerUsername)
			if hex.EncodeToString(aad) != tc.aadHex {
				t.Fatal("metadata AAD does not match shared fixture")
			}

			block, err := aes.NewCipher(key)
			if err != nil {
				t.Fatal(err)
			}
			gcm, err := cipher.NewGCM(block)
			if err != nil {
				t.Fatal(err)
			}
			gotCT := gcm.Seal(nil, nonce, []byte(tc.plaintext), aad)
			if hex.EncodeToString(gotCT) != tc.cipherHex {
				t.Fatal("metadata ciphertext does not match shared fixture")
			}

			combined := append(append([]byte{}, nonce...), wantCT...)
			got, err := DecryptGCMWithAAD(combined, key, aad)
			if err != nil {
				t.Fatal(err)
			}
			if !bytes.Equal(got, []byte(tc.plaintext)) {
				t.Fatal("metadata plaintext does not match shared fixture")
			}
		})
	}
}
